### PR TITLE
Add Mount RC failsafe handling

### PR DESCRIPTION
**Pull request by peterbarker**
@Hwurzburg

### CODE REVIEW

**Comment by peterbarker** [comment]
Maybe `s/in_rc_failsafe/in_failsafe/`

File: libraries/AP_Mount/AP_Mount.h [hunk 1]
```diff
@@ -106,6 +106,10 @@ class AP_Mount
     enum MAV_MOUNT_MODE get_mode() const { return get_mode(_primary); }
     enum MAV_MOUNT_MODE get_mode(uint8_t instance) const;
 
+    // get_rc_invalid_mode - returns rc invalid pointing coordinates of mount (i.e. no change, Retracted, Neutral)
```
Comment (Hwurzburg) [inline libraries/AP_Mount/AP_Mount.h:109]: appears you did everything! but still need to get rid of this, right?

**Comment by Hwurzburg** [comment]
want me to SITL and bench test this? and close mine?

File: libraries/AP_Mount/AP_Mount.h (hunk 1 above)
Comment (peterbarker) [reply libraries/AP_Mount/AP_Mount.h:109]: Done!

**Comment by peterbarker** [comment]
> want me to SITL and bench test this? and close mine?

Yes please!

File: libraries/AP_Mount/AP_Mount_Servo.cpp [hunk 2]
```diff
@@ -61,8 +61,22 @@ void AP_Mount_Servo::update()
...
+                       break;
+                    case MAV_MOUNT_MODE_NEUTRAL:
+                       _angle_bf_output_deg = _state._neutral_angles.get();
+                       break;
+                    default:
+                        //do nothing
+                        break;
+                    }
+            } else if (rc().has_valid_input()) {
             update_targets_from_rc();
```
Comment (khancyr) [inline libraries/AP_Mount/AP_Mount_Servo.cpp:78]: looks like is miss some indent here

File: libraries/AP_Mount/AP_Mount_Servo.cpp [hunk 3]
```diff
@@ -61,8 +61,22 @@ void AP_Mount_Servo::update()
...
         {
-            // update targets using pilot's rc inputs
+            // update targets using pilot's rc inputs or go to neutral or retracted targets if no rc
+            if (rc().in_rc_failsafe()) {
+                switch (_state._rcinvalid_mode.get()) {
+                    case MAV_MOUNT_MODE_RETRACT:
+                       _angle_bf_output_deg = _state._retract_angles.get();
+                       break;
+                    case MAV_MOUNT_MODE_NEUTRAL:
+                       _angle_bf_output_deg = _state._neutral_angles.get();
```
Comment (khancyr) [inline libraries/AP_Mount/AP_Mount_Servo.cpp:71]: looks like we can merge the two cases

**Review by khancyr — APPROVED** [review]
Looks good
It would be cool to write a small test for this too

**Comment by peterbarker** [comment]
On Fri, 8 Jan 2021, Pierre Kancir wrote:

> looks like we can merge the two cases

Read a little harder.  They *look* very similar :-)

**Comment by peterbarker** [comment]
On Fri, 8 Jan 2021, Pierre Kancir wrote:

> looks like is miss some indent here

Yeah, I retained the original indent for no particularly good reason.

**Comment by khancyr** [comment]
One word : dyslexia ;-S

File: libraries/AP_Mount/AP_Mount.cpp [hunk 4]
```diff
@@ -392,8 +392,22 @@ const AP_Param::GroupInfo AP_Mount::var_info[] = {
...
+    // @User: Standard
+    AP_GROUPINFO("2_RCINVLD_MODE", 43, AP_Mount, state[1]._rcinvalid_mode, -1),
 #endif // AP_MOUNT_MAX_INSTANCES > 1
 
+    // @Param: _RCINVLD_MODE
+    // @DisplayName: Mount RC Invalid Point Position
+    // @Description: Uses the designated modes target position if RC invalid in RC Targeting mode
+    // @Values: -1:No Change,0:Retracted,1:Neutral
+    // @User: Standard
+    AP_GROUPINFO("_RCINVLD_MODE", 44, AP_Mount, state[0]._rcinvalid_mode, -1),
```
Comment (rmackay9) [inline libraries/AP_Mount/AP_Mount.cpp:409]: maybe let's name this parameter MNT_FS_RC_ACTION

File: libraries/RC_Channel/RC_Channel.h [hunk 5]
```diff
@@ -380,7 +380,7 @@ class RC_Channels {
     void reset_mode_switch();
     virtual void read_mode_switch();
 
-    // has_valid_input should be pure-virtual when Plane is converted
+    virtual bool in_rc_failsafe() const { return true; };
```
Comment (tridge) [inline libraries/RC_Channel/RC_Channel.h:383]: can this be pure virtual?

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by peterbarker** [comment]
I tried the pure-virtual thing.  It didn't work out.  Too much stuff to be reverted when we move the failsafe handling up.

**Comment by Hwurzburg** [comment]
@peterbarker having a problem running SITL on this...if I use master/realfight griffin to get the mount, setting SIM_RC_FAIL puts it into failsafe...if I load this, it just disconnects RC from the controller box...maybe its a flightaxis thing...but cant test in STIL as is

**Comment by Hwurzburg** [comment]
btw, noticed that if I use RC mavproxy command it overrides the flight axis control box until I set RC X 0....probably should document...I assume it also does this in pure Mavproxy if joystick module is used?

File: libraries/AP_Mount/AP_Mount.cpp (hunk 4 above)
Comment (peterbarker) [reply libraries/AP_Mount/AP_Mount.cpp]: This was done.

File: libraries/RC_Channel/RC_Channel.h (hunk 5 above)
Comment (peterbarker) [reply libraries/RC_Channel/RC_Channel.h:399]: I tried the pure-virtual thing. It didn't work out. Too much stuff to be reverted when we move the failsafe handling up.

**Comment by peterbarker** [comment]
Closes https://github.com/ArduPilot/ardupilot/issues/24093

**Review by rmackay9 — APPROVED** [review]
This looks OK to me although I wonder if we really need the extra parameter.  I would have though just an MNTx_OPTION bit to move to Neutral on RC failsafe would have been enough.

**Comment by rmackay9** [comment]
We decided to use an option bit instead of adding the extra parameter, thanks!

**Comment by peterbarker** [comment]
```
Board                    AP_Periph  blimp  bootloader  copter  heli  iofirmware  plane  rover  sub
CubeOrange-periph-heavy  *                                                                     
CubeRedPrimary                      0      *           -256    -256              -256   -256   -256
Durandal                            *      *           -304    -304              -312   -304   -312
Hitec-Airspeed           *                                                                     
KakuteH7-bdshot                     *      *           -216    -216              -216   -216   -216
MatekF405                           *      *           *       *                 *      *      *
Pixhawk1-1M-bdshot                  *                  *       *                 *      *      *
f103-QiotekPeriph        *                                                                     
f303-Universal           *                                                                     
iomcu                                                                *                         
revo-mini                           *      *           *       *                 *      *      *
skyviper-journey                                       -112                                    
skyviper-v2450                                         *                                       
```

File: libraries/AP_Mount/AP_Mount_Backend.cpp [hunk 6]
```diff
@@ -66,6 +66,29 @@ bool AP_Mount_Backend::set_mode(MAV_MOUNT_MODE mode)
     return true;
 }
 
+// called when mount mode is RC-targetting, updates the mnt_target object from RC inputs:
+void AP_Mount_Backend::update_mnt_target_from_rc_target()
+{
+    if (rc().in_rc_failsafe()) {
+        if (option_set(Options::NEUTRAL_ON_RC_FS)) {
```
Comment (IamPete1) [inline libraries/AP_Mount/AP_Mount_Backend.cpp:73]: Needs param docs for new options.

**Review by IamPete1 — DISMISSED** [review]

File: libraries/AP_Mount/AP_Mount_Backend.cpp (hunk 6 above)
Comment (peterbarker) [reply libraries/AP_Mount/AP_Mount_Backend.cpp:73]: Fixed, thanks

**Review by rmackay9 — APPROVED** [review]

**Pull request merged**